### ISSUE TITLE
Descent MIDI notes don't release

### ISSUE THREAD

**Issue by mmontag** [Wohlstand/libADLMIDI#223]
Hi, a little mystery here, wonder if you can help debug. 

See this file: 
https://mmontag.github.io/chip-player-js/browse/Game%20MIDI/Descent%20(FM)?play=Game%20MIDI%2FDescent%20(FM)%2FGAME03%5Bint%5D.MID

Notice the synth bass loop that starts playing around 7.5 seconds. The notes are sustaining way too long, as compared to the sound in game: 
https://sectorgame.com/descentfiles/uploads/Music%20%26%20Sound%20Effects/Descent/Descent%20Midi%20Soundtrack%20%28OPL3%29/08%20%28Level%203%29%20Lunar%20Military%20Base.mp3

chip-player-js plays OPL3 through libADLMIDI realtime commands, and also loads the appropriate bank (HMI Descent: Int). My MIDI playback code is [here](https://github.com/mmontag/chip-player-js/blob/master/tinysoundfont/tinyplayer.c).

I checked the MIDI file and the notes are not held like that. 
I can't figure out if the problem is in the bank file or somewhere in the MIDI file -> sequencing -> synthesis pipeline.

**Comment by Wohlstand** [comment]
I have checked that by myself, and it's not a defect of player, it's the instrument itself, and it's normal behaviour :thinking: 
I have made a quick demo of that: https://www.youtube.com/watch?v=8RUz-dLZcIg
A bit later I'll give some random walkthrough video from the game itself where music is sounds...

**Comment by Wohlstand** [comment]
Or easier thing: try to choose absolutely any other bank even it will sound broken, but pay an attention on vibraphone (which a place of bass in that descent song), and you'll hear it's fine, not held :fox_face:  :wink:

**Comment by mmontag** [comment]
Thanks! 
Can you open the “vibraphone” in a bank editor and see the envelope params of the instrument?

I am guessing it has a long release time. So then, why do the notes get cut off early within the game?

**Comment by mmontag** [comment]
I just saw the demo video, is this your bank editor? looks awesome!

**Comment by Wohlstand** [comment]
You can open that in `fm_banks/bnk_files/intmelo.bnk` which is `Descent::Int` bank
![Снимок экрана от 2019-06-10 19-37-47](https://user-images.githubusercontent.com/6751442/59211013-864d4e00-8bb7-11e9-88c2-e31bcc82a414.png)

Yeah, everything I have showcased on the video with playback of the voice while I changing the instrument and the demo of playing the same song with a modified bank loaded as custom external WOPL bank file.

**Comment by Wohlstand** [comment]
P.S. @mmontag , I have checked out some MIDIs at your archive, and I see some interesting things! :fox_face: 
Myself I have a huge collection of MIDI (~21k files) include those I made by myself, I would exchange with some things, and I have to add into this archive from my collection (for example, I did my own full rip of Corridor 7 songs from IMF to MIDI conversion which is more accurate, and some other songs).

**Comment by mmontag** [comment]
@Wohlstand are you suggesting that you can contribute some MIDI files? If so, that would be great :) You can email me directly, matt.montag / gmaiI.

Also yea, libOPNMIDI sounds cool...and definitely better than the 64 MB soundfonts that try to do the same thing.

**Comment by Wohlstand** [comment]
@mmontag , at morning I have sent you an email, feel free to read it! :fox_face: :wink:

**Comment by mmontag** [comment]
footnote: 
All of this (custom banks plus quirky MIDI playback behavior) makes Descent music a good candidate for doing VGM rips from DosBox.

**Comment by jpcima** [comment]
I dont think the source code of the Descent FM MIDI driver was released unfortunately, which would permit to reproduce these tracks exactly.

Apparently the PC sources support the GM tracks only. When we have implemented libADLMIDI support in dxx-rebirth we have also used the GM tracks.
Since the Descent bank was arranged in different fashion than GM, I have just picked another from the set and used it as a default.

Note it can be bugs in original game engines also, which explain some differences of playback.

**Comment by mmontag** [comment]
Thanks for shedding some light on that @jpcima. I noticed some commented music stuff when browsing the Descent source, I guess it was a licensed product. dxx-rebirth is an awesome project!

**Comment by Wohlstand** [comment]
Okay, just now I have figured out for the sounding and I may guess two things I wanna check:
- Verify, are these BNK files validly interpreted as HMI banks? (parsing of HMI banks was done on the side of gen_adldata and OPL3-BE, and the result is same, but, looks isn't valid as some bytes may aren't matching here...)
- Get the way to extract banks from the game itself and compare to banks are pre-included with ADLMIDI

**Comment by Wohlstand** [comment]
Okay, I found the true reason for this difference:

[bass-test-SoS.zip](https://github.com/Wohlstand/libADLMIDI/files/5171144/bass-test-SoS.zip)

- libADLMIDI plays this bass with regular channel management: it doesn't hurt releasing notes, letting them complete self.
- HMI SOS (the library that was used in Descent) mutes the releasing stream. I bet, if I'll convert this into CMF, the sound will be the same :thinking: 

Also, I had to include my experimental set.

You can try to tweak this to compare things. Edit "setup.ini" to change names of test MIDI file and bank names, then run "setup.exe" and "midi test" to test the playback. Note that this thing won't play "hmp" files: hmi only.

**Comment by Wohlstand** [comment]
Okay, just now I finished some work to:
* Make an approximated "HMI" volume model (I didn't research the actual yet because of current reverse engineering challenges)
* When the HMI volume model is in use, make immediate channel re-using like HMI actually does, therefore Descent songs will sound closer to the game itself.

One note: while playing Descent songs, please enable the panning stereo mode (which is off by default). HMI does simulation of panned stereo by using of two channels in parallel.

**Comment by Wohlstand** [comment]
@mmontag, please apply the latest master state of libADLMIDI to you and verify, how Descent songs will be played on your end.

**Comment by mmontag** [comment]
@Wohlstand I'll try this out and let you know.

**Comment by mmontag** [comment]
@Wohlstand hm; I'm sad to say the track sounds the same as it did before, with the _HMI (Descent:: Int)_ bank. I've merged in the latest libADLMIDI master branch all the way up to commit 5703f4b (Sun Sep 6). Maybe something else I can try...? 
BTW I believe I've already enabled the panning mode:
```
  g_adlSynth = adl_init(sampleRate);
  adl_setSoftPanEnabled(g_adlSynth, 1);
  adl_setNumChips(g_adlSynth, 4);
```

**Comment by Wohlstand** [comment]
You need to rebuild your local adldata. If you did a set of WOPL files, please edit them with latest OPL3-BE and set them the HMI volume model. A quick fix: set volume model by libADLMIDI's API.

**Comment by Wohlstand** [comment]
The feature works if volume model is HMI. When other volume models used, then Descent songs will sound same as in past.

**Comment by Wohlstand** [comment]
Also: if you have the master copy, set volume model to AUTO, then HMI volume model should work for all banks of HMI type.

**Comment by mmontag** [comment]
Ok, I've added the AUTO volume model switch. It still sounds different from the MP3 I linked in the first post. I'm using adldata.cpp that is tracked in the repository - that should be up to date, right? 

It's deployed live now, in case you want to hear for yourself: https://mmontag.github.io/chip-player-js/browse/Game%20MIDI/Descent%20(FM)?play=Game%20MIDI%2FDescent%20(FM)%2FGAME03%5Bint%5D.MID

**Comment by Wohlstand** [comment]
> that should be up to date, right?

That fine. Re-genration is needed if you plan to pass your own set of banks.

> It still sounds different from the MP3 I linked in the first post.

The reason is that SOS is mote aggressive on channels allocation. Also, because SOS does use of full panned stereo by itself, it limits 9 channels only. If you will remove all tracks but bass, it will sound same in both SOS and libADLMIDI. I found a copy of SOS libs at Corridor 8 package, and I built a simple player with OpenWatcom to try various experiments. At least, other songs sounding correct without of making a mess of releasing notes.

**Comment by Wohlstand** [comment]
SOS - HMI Sound Operating System, a sound library used by Descent game.

**Comment by mmontag** [comment]
So this relies on a quirk of the sound engine... so maybe this should be addressed in the MIDI conversion process. It sounds very difficult to fix unless you have a dedicated HMI SOS emulation mode in libADLMIDI engine...

**Comment by Wohlstand** [comment]
I crafted another demo: 
- I left bass play solo: it will play like now at libADLMIDI
- Then, I play drums and bass seems gets being shouted up by each small drum instrument :thinking: 
[GAME03z-sos-demo.zip](https://github.com/Wohlstand/libADLMIDI/files/5180536/GAME03z-sos-demo.zip)
I will need to tune a bit the algorithm that manages OPL3 channels and make it shut up any releasing voices, not only "namesakes" (which those are playing the same instrument)

**Comment by Wohlstand** [comment]
@mmontag , just now I made a small tweak, and seems, HMI behavior is now more accurate :fox_face:

**Comment by Wohlstand** [comment]
P.S. Please pull the top master, I did one minor mistake that breaks all other models, I fixed that.

**Comment by Wohlstand** [comment]
A quick demo: [adl[L]-vs-sos[R].ogg.zip](https://github.com/Wohlstand/libADLMIDI/files/5186982/adl.L.-vs-sos.R.ogg.zip)
libADLMIDI at left, SOSPLAY at right. Both ran in DosBox to make a fair comparison.

**Comment by mmontag** [comment]
Yes, it sounds great now! I am getting the same result as your example audio. Very impressive :)
I hope your effort yields more accurate playback in other music as well.

**Comment by Wohlstand** [comment]
>  I hope your effort yields more accurate playback in other music as well.

Working hard on this :) 
Also, my recent works I did to make accurate sounding of AIL and DMX banks: I had to replicate bugs of DMX with AM voices, and make the sounding of volume model same as in Doom games, and AIL volume model was been tuned. I had to inspect ASM source of AIL2, released by Miles in 2000. HMI is hardest thing because I had to dig for tools the universe and hack by disassemblers :zany_face: . Also, I had to replicate the bug of AM voices for Apogee Sound System model, especially this makes accurate sounding of FM music of Blood game.

**Comment by Wohlstand** [comment]
I guess, the task is solved, I'll close it! :fox_face:

**Comment by Wohlstand** [comment]
Thank you @mmontag for verifying and confirming the work! :)

**Comment by Wohlstand** [comment]
btw, one bug I will need to fix #233, otherwise, Doom's bank makes some odd noise while playing some songs...

EDIT: Fixed
